scene/3d/mesh_instance.cpp

```diff
@@ -163,15 +173,280 @@ void MeshInstance::_resolve_skeleton_path() {
+
+			visual_server->instance_attach_skeleton(get_instance(), skin_ref->get_skeleton());
+
+			if (!mesh_instance.is_null()) {
+				mesh_instance = Ref<Mesh>();
+				software_skinning_data.clear();
+				update_mesh = true;
+			}
+		}
 	} else {
-		VisualServer::get_singleton()->instance_attach_skeleton(get_instance(), RID());
+		visual_server->instance_attach_skeleton(get_instance(), RID());
+		if (!mesh_instance.is_null()) {
+			mesh_instance = Ref<Mesh>();
+			software_skinning_data.clear();
+			update_mesh = true;
+		}
+	}
+
+	RID render_mesh = mesh_instance.is_valid() ? mesh_instance->get_rid() : mesh->get_rid();
+	set_base(render_mesh);
+
+	if (update_mesh) {
+		// Update instance materials after switching mesh
+		ERR_FAIL_COND(mesh.is_null());
+		int surface_count = mesh->get_surface_count();
+		for (int surface_index = 0; surface_index < surface_count; ++surface_index) {
+			if (materials[surface_index].is_valid()) {
+				visual_server->instance_set_surface_material(get_instance(), surface_index, materials[surface_index]->get_rid());
+			}
+		}
 	}
 }
 
+void MeshInstance::_update_skinning() {
+	ERR_FAIL_COND(!_is_software_skinning_enabled());
+	ERR_FAIL_COND(!is_visible());
+
+	ERR_FAIL_COND(skin_ref.is_null());
+	RID skeleton = skin_ref->get_skeleton();
+	ERR_FAIL_COND(!skeleton.is_valid());
+
+	VisualServer *visual_server = VisualServer::get_singleton();
+
+	// Prepare bone transforms.
+	int num_bones = visual_server->skeleton_get_bone_count(skeleton);
+	ERR_FAIL_COND(num_bones <= 0);
+	Transform *bone_transforms = (Transform *)alloca(sizeof(Transform) * num_bones);
+	for (int bone_index = 0; bone_index < num_bones; ++bone_index) {
+		bone_transforms[bone_index] = visual_server->skeleton_bone_get_transform(skeleton, bone_index);
+	}
+
+	ERR_FAIL_COND(!mesh_instance.is_valid());
+	RID mesh_rid = mesh_instance->get_rid();
+	ERR_FAIL_COND(!mesh_rid.is_valid());
+
+	// Apply skinning.
+	int surface_count = mesh_instance->get_surface_count();
+	for (int surface_index = 0; surface_index < surface_count; ++surface_index) {
+		uint32_t format = mesh_instance->surface_get_format(surface_index);
+
+		ERR_CONTINUE(0 == (format & Mesh::ARRAY_FORMAT_BONES));
+		ERR_CONTINUE(0 == (format & Mesh::ARRAY_FORMAT_WEIGHTS));
+
+		const int vertex_count = mesh_instance->surface_get_array_len(surface_index);
+		const int index_count = mesh_instance->surface_get_array_index_len(surface_index);
+
+		uint32_t array_offsets[Mesh::ARRAY_MAX];
+		uint32_t stride = visual_server->mesh_surface_make_offsets_from_format(format, vertex_count, index_count, array_offsets);
+		uint32_t offset_vertices = array_offsets[Mesh::ARRAY_VERTEX];
+		uint32_t offset_normals = array_offsets[Mesh::ARRAY_NORMAL];
+		uint32_t offset_bones = array_offsets[Mesh::ARRAY_BONES];
+		uint32_t offset_weights = array_offsets[Mesh::ARRAY_WEIGHTS];
+
+		PoolByteArray buffer = visual_server->mesh_surface_get_array(mesh_rid, surface_index);
+		PoolByteArray::Read buffer_read = buffer.read();
+
```

Comment:
In both cases can we not read and write to the same buffer? If you are using the local buffer, then it may well be more expensive to jump around in cache between 2 buffers needlessly.